File: frontend/src/api/unityRunner.ts
import apiClient from './client';
import type { TestSession } from './sessions';
import type { MetricCatalog, TestScope, TestScopeSummary } from '@/lib/testScope';

export interface UnityEngineResource {
  id: string;
  name: string;
  version: string;
  executable_path: string;
  enabled: boolean;
  is_default: boolean;
  exists: boolean;
  notes?: string | null;
}

export interface UnitySceneResource {
  id: string;
  name: string;
  description: string | null;
  project_path: string;
  scene_path: string;
  scene_file_path: string;
  enabled: boolean;
  is_default: boolean;
  exists: boolean;
  collector_package_name: string;
  collector_package_path: string | null;
  manifest_has_plugin: boolean;
  tags: string[];
}

export interface UnityQualityChecks {
  lighting: boolean;
  materials: boolean;
  post_processing: boolean;
  physics: boolean;
}

export interface UnityMetricChecks {
  frame_rate: boolean;
  frame_time: boolean;
  cpu: boolean;
  gpu: boolean;
  memory: boolean;
  device_info: boolean;
}

export interface UnityTestStartRequest {
  project_id: number;
  unity_engine_id: string;
  scene_resource_id: string;
  test_scope?: TestScope;
  quality_checks: UnityQualityChecks;
  quality_metric_checks?: Record<string, boolean>;
  metric_checks?: Partial<UnityMetricChecks>;
  collect_interval: number;
  frame_rate_duration_seconds: number;
  metrics_duration_seconds: number;
  batchmode: boolean;
  ensure_plugin: boolean;
}

export interface UnityTestTask {
  id: number;
  name: string;
  description: string | null;
  status: string;
  task_type: string;
  project_id: number | null;
  scene_id: number | null;
  config: Record<string, unknown> | null;
  started_at: string | null;
  completed_at: string | null;
  duration_seconds: number | null;
  error_message: string | null;
}

export interface UnityTestStartResponse {
  task: UnityTestTask;
  session: TestSession;
  engine: UnityEngineResource;
  scene: UnitySceneResource;
  process_id: number | null;
  task_config_path: string;
  unity_log_path: string;
  runner_log_path: string;
}

export interface UnityTestStopResponse {
  task: UnityTestTask;
  session: TestSession | null;
}

export interface UnityTaskLogsResponse {
  task: UnityTestTask;
  session: TestSession | null;
  runner_log_path: string | null;
  unity_log_path: string | null;
  lines: string[];
}

export interface UnityRealtimeProgress {
  type: 'unity_progress';
  task_id: number;
  session_id: number;
  phase: string;
  phase_label: string;
  progress: number;
  remaining_seconds: number;
  sample_count: number;
  fps: number;
  frame_time_ms: number;
  raw_frame_time_ms: number;
  cpu_usage_percent: number;
  gpu_usage_percent: number;
  memory_mb: number;
  managed_memory_mb: number;
  graphics_memory_mb: number;
  system_memory_mb: number;
  draw_calls: number;
  triangles: number;
  vertices: number;
  active_light_count: number;
  realtime_light_count: number;
  shadow_caster_count: number;
  reflection_probe_count: number;
  material_count: number;
  unique_material_count: number;
  transparent_material_count: number;
  post_process_volume_count: number;
  render_texture_count: number;
  rigidbody_count: number;
  collider_count: number;
  is_xr_active: boolean;
  xr_device_name: string;
  device_model: string;
  operating_system: string;
  unity_version: string;
  graphics_device_name: string;
  render_pipeline: string;
  screen_resolution: string;
  received_at?: string;
  test_scope_version?: number;
  test_scope_summary?: TestScopeSummary;
  selected_metric_ids?: string[];
  skipped_metric_ids?: string[];
}

export function createUnityProgressWebSocket(taskId: number): WebSocket {
  const apiBase = import.meta.env.VITE_API_BASE_URL || '/api/v1';
  const baseUrl = new URL(apiBase, window.location.origin);
  baseUrl.protocol = baseUrl.protocol === 'https:' ? 'wss:' : 'ws:';
  baseUrl.pathname = `${baseUrl.pathname.replace(/\/$/, '')}/unity-runner/progress/${taskId}/ws`;
  baseUrl.search = `token=${encodeURIComponent(localStorage.getItem('xr_token') || '')}`;
  return new WebSocket(baseUrl);
}

export const unityRunnerApi = {
  getTestMetricsCatalog: async (): Promise<MetricCatalog> => {
    const response = await apiClient.get<MetricCatalog>('/unity-runner/test-metrics/catalog');
    return response.data;
  },

  getDefaultTestScope: async (): Promise<{ default_scope: TestScope; scope_summary: TestScopeSummary }> => {
    const response = await apiClient.get<{ default_scope: TestScope; scope_summary: TestScopeSummary }>(
      '/unity-runner/test-metrics/default-scope',
    );
    return response.data;
  },

  listEngines: async (): Promise<UnityEngineResource[]> => {
    const response = await apiClient.get<{ items: UnityEngineResource[] }>('/unity-runner/engines');
    return response.data.items;
  },

  listScenes: async (params?: { project_id?: number }): Promise<UnitySceneResource[]> => {
    const response = await apiClient.get<{ items: UnitySceneResource[] }>('/unity-runner/scenes', { params });
    return response.data.items;
  },

  startTest: async (data: UnityTestStartRequest): Promise<UnityTestStartResponse> => {
    const response = await apiClient.post<UnityTestStartResponse>('/unity-runner/test-tasks/start', data);
    return response.data;
  },

  stopTest: async (taskId: number): Promise<UnityTestStopResponse> => {
    const response = await apiClient.post<UnityTestStopResponse>(`/unity-runner/test-tasks/${taskId}/stop`);
    return response.data;
  },

  getTaskLogs: async (taskId: number, params?: { tail_lines?: number }): Promise<UnityTaskLogsResponse> => {
    const response = await apiClient.get<UnityTaskLogsResponse>(`/unity-runner/test-tasks/${taskId}/logs`, { params });
    return response.data;
  },

  getLatestProgress: async (taskId: number): Promise<UnityRealtimeProgress | null> => {
    const response = await apiClient.get<{ item: UnityRealtimeProgress | null }>(`/unity-runner/progress/${taskId}/latest`);
    return response.data.item;
  },
};

File: backend/app/routers/unity_runner.py
from typing import Any, Optional

from fastapi import APIRouter, Depends, Query, Request
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.core.permissions import Permission, require_permission
from app.database import get_db
from app.models.user import User
from app.services.audit_service import log_audit
from app.services.unity_runner_service import UnityRunnerService


router = APIRouter(prefix="/unity-runner", tags=["Unity 本地测试"])


class QualityChecks(BaseModel):
    lighting: bool = True
    materials: bool = True
    post_processing: bool = True
    physics: bool = True


class MetricChecks(BaseModel):
    frame_rate: bool = True
    frame_time: bool = True
    cpu: bool = True
    gpu: bool = True
    memory: bool = True
    device_info: bool = True


class TestScopeRequest(BaseModel):
    schema_version: int = Field(default=1, ge=1)
    source: str | None = None
    basic_metrics: dict[str, bool] = Field(default_factory=dict)
    quality_categories: dict[str, bool] = Field(default_factory=dict)
    quality_metrics: dict[str, bool] = Field(default_factory=dict)


class UnityTestStartRequest(BaseModel):
    project_id: int = Field(..., gt=0)
    unity_engine_id: str = Field(..., min_length=1)
    scene_resource_id: str = Field(..., min_length=1)
    test_scope: TestScopeRequest | None = None
    quality_checks: QualityChecks = Field(default_factory=QualityChecks)
    quality_metric_checks: dict[str, bool] = Field(default_factory=dict)
    metric_checks: MetricChecks = Field(default_factory=MetricChecks)
    collect_interval: float = Field(1.0, ge=0.1, le=10.0)
    frame_rate_duration_seconds: float = Field(30.0, ge=1.0, le=600.0)
    metrics_duration_seconds: float = Field(30.0, ge=1.0, le=600.0)
    batchmode: bool = False
    ensure_plugin: bool = True


@router.get("/test-metrics/catalog")
async def get_test_metrics_catalog(
    current_user: User = Depends(require_permission(Permission.TEST_VIEW)),
    db: Session = Depends(get_db),
):
    return UnityRunnerService(db).get_test_metrics_catalog()


@router.get("/test-metrics/default-scope")
async def get_default_test_scope_for_run(
    current_user: User = Depends(require_permission(Permission.TEST_VIEW)),
    db: Session = Depends(get_db),
):
    scope = UnityRunnerService(db).get_default_test_scope_for_run()
    from app.services.test_scope_service import TestScopeService

    return {
        "default_scope": scope,
        "scope_summary": TestScopeService.build_scope_summary(scope),
    }


@router.get("/engines")
async def list_unity_engines(
    current_user: User = Depends(require_permission(Permission.TEST_VIEW)),
    db: Session = Depends(get_db),
):
    service = UnityRunnerService(db)
    return {"items": service.list_engines()}


@router.get("/scenes")
async def list_unity_scene_resources(
    project_id: Optional[int] = Query(None, gt=0),
    current_user: User = Depends(require_permission(Permission.TEST_VIEW)),
    db: Session = Depends(get_db),
):
    service = UnityRunnerService(db)
    return {"items": service.list_scenes(project_id=project_id)}


@router.post("/test-tasks/start")
async def start_unity_test_task(
    request: Request,
    payload: UnityTestStartRequest,
    current_user: User = Depends(require_permission(Permission.TEST_EXECUTE)),
    db: Session = Depends(get_db),
):
    service = UnityRunnerService(db)
    result = service.start_test(
        project_id=payload.project_id,
        unity_engine_id=payload.unity_engine_id,
        scene_resource_id=payload.scene_resource_id,
        test_scope=payload.test_scope.model_dump() if payload.test_scope else None,
        quality_checks=payload.quality_checks.model_dump(),
        quality_metric_checks=payload.quality_metric_checks,
        metric_checks=payload.metric_checks.model_dump(),
        collect_interval=payload.collect_interval,
        frame_rate_duration_seconds=payload.frame_rate_duration_seconds,
        metrics_duration_seconds=payload.metrics_duration_seconds,
        batchmode=payload.batchmode,
        ensure_plugin=payload.ensure_plugin,
        creator_id=current_user.id,
    )

    await log_audit(
        db=db,
        action="test_execute",
        user_id=current_user.id,
        username=current_user.username,
        ip_address=request.client.host if request.client else None,
        resource_type="test_task",
        resource_id=str(result["task"]["id"]),
        details={
            "source": "web_unity_runner",
            "project_id": payload.project_id,
            "session_id": result["session"]["id"],
            "unity_engine_id": payload.unity_engine_id,
            "scene_resource_id": payload.scene_resource_id,
        },
    )

    return result


@router.post("/test-tasks/{task_id}/stop")
async def stop_unity_test_task(
    request: Request,
    task_id: int,
    current_user: User = Depends(require_permission(Permission.TEST_EXECUTE)),
    db: Session = Depends(get_db),
):
    service = UnityRunnerService(db)
    result = service.stop_test(task_id)

    await log_audit(
        db=db,
        action="test_stop",
        user_id=current_user.id,
        username=current_user.username,
        ip_address=request.client.host if request.client else None,
        resource_type="test_task",
        resource_id=str(task_id),
        details={"source": "web_unity_runner"},
    )

    return result


@router.get("/test-tasks/{task_id}/logs")
async def get_unity_test_task_logs(
    task_id: int,
    tail_lines: int = Query(400, ge=20, le=1000),
    current_user: User = Depends(require_permission(Permission.TEST_VIEW)),
    db: Session = Depends(get_db),
):
    service = UnityRunnerService(db)
    return service.get_task_logs(task_id, tail_lines=tail_lines)
